cf-net/cf-net.c

```diff
@@ -0,0 +1,766 @@
+    return SocketConnect(server, port, 100, true,
+                         txtaddr, sizeof(txtaddr));
+}
+
+//*******************************************************************
+// COMMANDS:
+//*******************************************************************
+
+static int CFNetHelpTopic(const char *topic)
+{
+    assert(topic != NULL);
+    bool found = false;
+    for (int i = 0; COMMANDS[i].name != NULL; ++i)
+    {
+        if (strcmp(COMMANDS[i].name, topic) == 0)
+        {
+            printf("Command:     %s\n",  COMMANDS[i].name);
+            printf("Usage:       %s\n",  COMMANDS[i].usage);
+            printf("Description: %s\n", COMMANDS[i].description);
+            found = true;
+            break;
+        }
+    }
+
+    // Add more detailed explanation here if necessary:
+    if (strcmp("help", topic) == 0)
+    {
+        printf("\nYou did it, you used the help command!\n");
+    }
+    else if (strcmp("stat", topic) == 0)
+    {
+        printf("\nFor security reasons the server doesn't give any additional"
+               "\ninformation if unsuccessful. 'Could not stat' message can"
+               "\nindicate wrong path, the file doesn't exist, permission "
+               "\ndenied, or something else.\n");
+    }
+    else if (strcmp("get", topic) == 0)
+    {
+        printf("\ncf-net get is comprised of two requests, STAT and GET."
+               "\nThe CFEngine GET protocol command requires a STAT first to"
+               "\ndetermine file size. By default the file is saved as its"
+               "\nbasename in current working directory (cwd). Override this"
+               "\nusing the -o filename option (-o - for stdout).\n");
+    }
+    else
+    {
+        if (found == false)
+        {
+            printf("Unknown help topic: '%s'\n", topic);
+            return 1;
+        }
+    }
+    return 0;
+}
+
+static int CFNetHelp(const char *topic)
+{
+    if (topic != NULL)
+    {
+        CFNetHelpTopic(topic);
+    }
+    else
+    {
+        Writer *w = FileWriter(stdout);
+        WriterWriteHelp(w, "cf-net", OPTIONS, HINTS, false, COMMANDS);
+        FileWriterDetach(w);
+        exit(EXIT_SUCCESS);
+    }
+    return 0;
+}
+
+static int CFNetConnectSingle(const char *server, bool print)
+{
+    AgentConnection *conn = CFNetOpenConnection(server);
+    if (conn == NULL)
+    {
+        return 1;
+    }
+    if (print == true)
+    {
+        printf("Connected & authenticated successfully to '%s'\n", server);
+    }
+    DisconnectServer(conn);
+    return 0;
+}
+
+static int CFNetConnect(CFNetOptions *opts, const char* hostname, char **args)
+{
+    if (hostname != NULL)
+    {
+        CFNetConnectSingle(hostname, true); // FIXME
+        return 0;
+    }
+    // TODO
+    if (args == NULL)
+    {
+        return 0;
+    }
+    if (opts->used_default)
+    {
+        return 0;
+    }
+    for (int i = 0; args[i] != NULL; ++i)
+    {
+        CFNetConnectSingle(args[i], true);
```

Comment:
I'm having some trouble understanding this function, and it has some TODOs and FIXMEs: Why are there two calls to `CFNetConnectSingle` here? And why does the second one take direct arguments from `args`? I thought the hostnames were supposed to be passed in via the `hostname` variable. Some clarification would help here.